File: backend/api/schemas/datasources.py
import re
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, SecretStr, field_validator

SCHEMA_NAME = re.compile(r"^[A-Za-z_][A-Za-z0-9_$]{0,62}$")


class PostgreSQLConnectionInput(BaseModel):
    source_type: Literal["postgresql"] = "postgresql"
    host: str = Field(min_length=1, max_length=255)
    port: int = Field(default=5432, ge=1, le=65535)
    database: str = Field(min_length=1, max_length=255)
    username: str = Field(min_length=1, max_length=255)
    password: SecretStr
    ssl_mode: Literal["disable", "prefer", "require", "verify-ca", "verify-full"] = "prefer"
    allowed_schemas: list[str] = Field(
        default_factory=lambda: ["public"], min_length=1, max_length=20
    )

    @field_validator("host", "database", "username")
    @classmethod
    def strip_required_value(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("Value must not be blank")
        return stripped

    @field_validator("allowed_schemas")
    @classmethod
    def validate_schemas(cls, value: list[str]) -> list[str]:
        normalized = list(dict.fromkeys(item.strip() for item in value))
        if not normalized or any(not SCHEMA_NAME.fullmatch(item) for item in normalized):
            raise ValueError("Allowed schemas must be valid PostgreSQL identifiers")
        return normalized


class DatasourceCreate(PostgreSQLConnectionInput):
    name: str = Field(min_length=1, max_length=120)

    @field_validator("name")
    @classmethod
    def strip_name(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("Connection name must not be blank")
        return stripped


class ConnectionTestResponse(BaseModel):
    status: Literal["ok"] = "ok"
    database: str
    server_version: str
    read_only_transaction: bool


class DatasourceSummary(BaseModel):
    id: UUID
    name: str
    source_type: str
    database_name: str
    safe_host: str
    port: int
    ssl_mode: str
    allowed_schemas: list[str]
    status: str
    is_active: bool
    entity_count: int
    relationship_count: int
    profile_count: int
    semantic_term_count: int
    metric_count: int
    embedding_count: int
    pii_excluded_count: int
    semantic_status: str
    semantic_error_code: str | None
    last_refreshed_at: datetime | None
    last_error_code: str | None
    created_at: datetime
    updated_at: datetime


class FieldSummary(BaseModel):
    id: UUID
    name: str
    native_type: str
    normalized_type: str
    nullable: bool
    ordinal: int
    primary_key: bool
    unique: bool
    description: str | None
    profile: dict[str, object] | None
    profile_sample_size: int | None
    profile_excluded: bool


class EntitySummary(BaseModel):
    id: UUID
    schema_name: str
    name: str
    entity_type: str
    description: str | None
    business_terms: list[str]
    metrics: list[str]
    fields: list[FieldSummary]


class RelationshipSummary(BaseModel):
    id: UUID
    name: str
    source: str
    target: str


class DatasourceDetail(DatasourceSummary):
    entities: list[EntitySummary]
    relationships: list[RelationshipSummary]


class OnboardingStatusResponse(BaseModel):
    datasource_id: UUID
    status: str
    entity_count: int
    relationship_count: int
    last_refreshed_at: datetime | None
    last_error_code: str | None


class SemanticManifestTerm(BaseModel):
    id: UUID
    term: str
    description: str | None
    confidence: float
    source: str


class SemanticManifestMetric(BaseModel):
    id: UUID
    name: str
    expression: str
    description: str | None
    confidence: float
    source: str
    verified: bool


class SemanticManifestField(BaseModel):
    id: UUID
    name: str
    native_type: str
    normalized_type: str
    nullable: bool
    ordinal: int
    primary_key: bool
    unique: bool
    description: str | None
    profile: dict[str, object] | None
    profile_sample_size: int | None
    profile_excluded: bool
    semantic_terms: list[SemanticManifestTerm]


class SemanticManifestEntity(BaseModel):
    id: UUID
    schema_name: str
    name: str
    entity_type: str
    description: str | None
    semantic_terms: list[SemanticManifestTerm]
    metrics: list[SemanticManifestMetric]
    embedding_artifact_id: UUID | None
    fields: list[SemanticManifestField]


class SemanticManifestRelationship(BaseModel):
    id: UUID
    source: str
    target: str
    source_field: str | None
    target_field: str | None
    relationship_type: str
    provenance: Literal["declared", "inferred"]
    confidence: float
    evidence: list[str]
    generation_eligible: bool


class SemanticManifestConfiguration(BaseModel):
    model_config_version: str
    retrieval_config_version: str
    generation_provider: str
    generation_model: str
    fallback_provider: str
    fallback_model: str
    embedding_model: str
    embedding_dimensions: int
    relationship_inferred_min_confidence: float
    privacy_policy_version: str
    skill_versions: dict[str, str]


class SemanticManifestResponse(BaseModel):
    manifest_id: UUID
    datasource_id: UUID
    datasource_name: str
    datasource_type: str
    version: int
    manifest_hash: str
    metadata_hash: str
    profile_hash: str
    configuration: SemanticManifestConfiguration
    entities: list[SemanticManifestEntity]
    relationships: list[SemanticManifestRelationship]
    created_at: datetime

File: frontend/lib/datasources.ts
import { apiRequest } from "@/lib/api-client";

export type DatasourceStatus = "draft" | "testing" | "introspecting" | "ready" | "failed";
export type SslMode = "disable" | "prefer" | "require" | "verify-ca" | "verify-full";

export interface DatasourceConnectionInput {
  source_type: "postgresql";
  host: string;
  port: number;
  database: string;
  username: string;
  password: string;
  ssl_mode: SslMode;
  allowed_schemas: string[];
}

export interface DatasourceCreateInput extends DatasourceConnectionInput {
  name: string;
}

export interface ConnectionTestResult {
  status: "ok";
  database: string;
  server_version: string;
  read_only_transaction: boolean;
}

export interface DatasourceSummary {
  id: string;
  name: string;
  source_type: string;
  database_name: string;
  safe_host: string;
  port: number;
  ssl_mode: string;
  allowed_schemas: string[];
  status: DatasourceStatus;
  is_active: boolean;
  entity_count: number;
  relationship_count: number;
  profile_count: number;
  semantic_term_count: number;
  metric_count: number;
  embedding_count: number;
  pii_excluded_count: number;
  semantic_status: "not_configured" | "configuration_required" | "indexing" | "ready" | "stale" | "failed";
  semantic_error_code: string | null;
  last_refreshed_at: string | null;
  last_error_code: string | null;
  created_at: string;
  updated_at: string;
}

export interface DatasourceField {
  id: string;
  name: string;
  native_type: string;
  normalized_type: string;
  nullable: boolean;
  ordinal: number;
  primary_key: boolean;
  unique: boolean;
  description: string | null;
  profile: Record<string, unknown> | null;
  profile_sample_size: number | null;
  profile_excluded: boolean;
}

export interface DatasourceEntity {
  id: string;
  schema_name: string;
  name: string;
  entity_type: string;
  description: string | null;
  business_terms: string[];
  metrics: string[];
  fields: DatasourceField[];
}

export interface DatasourceRelationship {
  id: string;
  name: string;
  source: string;
  target: string;
  source_field: string | null;
  target_field: string | null;
  relationship_type: string;
  provenance: "declared" | "inferred";
  confidence: number;
  evidence: string[];
  generation_eligible: boolean;
}

export interface DatasourceDetail extends DatasourceSummary {
  entities: DatasourceEntity[];
  relationships: DatasourceRelationship[];
}

export interface SemanticManifestTerm {
  id: string;
  term: string;
  description: string | null;
  confidence: number;
  source: string;
}

export interface SemanticManifestMetric {
  id: string;
  name: string;
  expression: string;
  description: string | null;
  confidence: number;
  source: string;
  verified: boolean;
}

export interface SemanticManifestField extends DatasourceField {
  semantic_terms: SemanticManifestTerm[];
}

export interface SemanticManifestEntity {
  id: string;
  schema_name: string;
  name: string;
  entity_type: string;
  description: string | null;
  semantic_terms: SemanticManifestTerm[];
  metrics: SemanticManifestMetric[];
  embedding_artifact_id: string | null;
  fields: SemanticManifestField[];
}

export interface SemanticManifestRelationship {
  id: string;
  source: string;
  target: string;
  source_field: string | null;
  target_field: string | null;
  relationship_type: string;
  provenance: "declared" | "inferred";
  confidence: number;
  evidence: string[];
  generation_eligible: boolean;
}

export interface SemanticManifest {
  manifest_id: string;
  datasource_id: string;
  datasource_name: string;
  datasource_type: string;
  version: number;
  manifest_hash: string;
  metadata_hash: string;
  profile_hash: string;
  configuration: {
    model_config_version: string;
    retrieval_config_version: string;
    generation_provider: string;
    generation_model: string;
    fallback_provider: string;
    fallback_model: string;
    embedding_model: string;
    embedding_dimensions: number;
    relationship_inferred_min_confidence: number;
    privacy_policy_version: string;
    skill_versions: Record<string, string>;
  };
  entities: SemanticManifestEntity[];
  relationships: SemanticManifestRelationship[];
  created_at: string;
}

export const listDatasources = () => apiRequest<DatasourceSummary[]>("/api/v1/datasources");

export function testDatasourceConnection(payload: DatasourceConnectionInput) {
  return apiRequest<ConnectionTestResult>("/api/v1/datasources/test", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function createDatasource(payload: DatasourceCreateInput) {
  return apiRequest<DatasourceDetail>("/api/v1/datasources", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export const getDatasource = (id: string) =>
  apiRequest<DatasourceDetail>(`/api/v1/datasources/${id}`);
export const activateDatasource = (id: string) =>
  apiRequest<DatasourceDetail>(`/api/v1/datasources/${id}/activate`, { method: "POST" });
export const refreshDatasource = (id: string) =>
  apiRequest<DatasourceDetail>(`/api/v1/datasources/${id}/refresh`, { method: "POST" });
export const getSemanticManifest = (id: string) =>
  apiRequest<SemanticManifest>(`/api/v1/datasources/${id}/semantic-manifest`);
